### PR TITLE
Add pydantic materializer

### PR DESCRIPTION
**Pull request by htahir1**
## Describe changes
I added a pydantic materializer. To check:

```
import random

from zenml.pipelines import pipeline
from zenml.steps import Output, step
from pydantic import BaseModel

class Lol(BaseModel):
    a: int = 1

@step(enable_cache=False)
def get_first_num() -> Lol:
    """Returns an integer."""

    return Lol(a=5)


@step(enable_cache=False)
def get_random_int() -> Output(random_num=int):
    """Get a random integer between 0 and 10"""
    return random.randint(0, 10)


@step
def subtract_numbers(first_num: int, lol: Lol) -> Output(result=int):
    """Subtract random_num from first_num."""
    print(lol.a)
    return first_num - lol.a


@pipeline
def hamzas_example_pipeline2(get_first_num, get_random_int, subtract_numbers):
    # Link all the steps artifacts together
    first_num = get_first_num()
    random_num = get_random_int()
    subtract_numbers(random_num, first_num )


if __name__ == "__main__":
    # Initialize a new pipeline run
    aep = hamzas_example_pipeline2(
        get_first_num=get_first_num(),
        get_random_int=get_random_int(),
        subtract_numbers=subtract_numbers(),
    )
    DAG = aep.run()
    
```

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/materializers/pydantic_materializer.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        contents = super().load(dict)
+        return data_type(**contents)
+
+    def save(self, data: BaseModel) -> None:
+        """Serialize a BaseModel to JSON.
+
+        Args:
+            data: The data to store.
+        """
+        super().save(data.dict())
```
Comment (schustmi) [inline src/zenml/materializers/pydantic_materializer.py:55]: This does not work unfortunately, as pydantic supports more types than the default json serializable ones (e.g. datetime, UUID).

We need to either
- store `data.json()` in a file (I don't see a reason not to do this)
- use `super().save(json.loads(data.json()))

Comment (htahir1) [reply src/zenml/materializers/pydantic_materializer.py:55]: But `data.json` basically does `data.dict` in the background right?

Comment (schustmi) [reply src/zenml/materializers/pydantic_materializer.py:55]: And in the process converts classes like UUID and datetime which the current solution won't cover. We have this pattern in our code a few times for exactly that reason, see e.g. https://github.com/zenml-io/zenml/blob/main/src/zenml/utils/pydantic_utils.py#L193-L194

Comment (htahir1) [reply src/zenml/materializers/pydantic_materializer.py]: Ah I got it. I now did it so that it goes via `data.json()`

**Review by schustmi — APPROVED** [review]
🦭

**Review by AlexejPenner — APPROVED** [review]
:seal:

**Pull request merged**